core/src/main/scala/org/apache/spark/storage/PushBasedFetchHelper.scala

```diff
@@ -0,0 +1,336 @@
+   *    See fetchLocalBlock.
+   * 2. There is a failure when fetching remote shuffle block chunks.
+   * 3. There is a failure when processing SuccessFetchResult which is for a shuffle chunk
+   *    (local or remote).
+   *
+   * @return number of blocks processed
+   */
+  def initiateFallbackBlockFetchForMergedBlock(
+      blockId: BlockId,
+      address: BlockManagerId): Int = {
+    assert(blockId.isInstanceOf[ShuffleBlockId] || blockId.isInstanceOf[ShuffleBlockChunkId])
+    logWarning(s"Falling back to fetch the original unmerged blocks for merged block $blockId")
+    // Increase the blocks processed since we will process another block in the next iteration of
+    // the while loop in ShuffleBlockFetcherIterator.next().
+    var blocksProcessed = 1
+    val fallbackBlocksByAddr: Iterator[(BlockManagerId, Seq[(BlockId, Long, Int)])] =
+      blockId match {
+        case shuffleBlockId: ShuffleBlockId =>
+          mapOutputTracker.getMapSizesForMergeResult(
+            shuffleBlockId.shuffleId, shuffleBlockId.reduceId)
+        case _ =>
+          val shuffleChunkId = blockId.asInstanceOf[ShuffleBlockChunkId]
+          val chunkBitmap: RoaringBitmap = chunksMetaMap.remove(shuffleChunkId)
+          assert(chunkBitmap != null)
+          // When there is a failure to fetch a remote merged shuffle block chunk, then we try to
+          // fallback not only for that particular remote shuffle block chunk but also for all the
+          // pending block chunks that belong to the same host. The reason for doing so is that it
+          // is very likely that the subsequent requests for merged block chunks from this host will
+          // fail as well. Since, push-based shuffle is best effort and we try not to increase the
+          // delay of the fetches, we immediately fallback for all the pending shuffle chunks in the
+          // fetchRequests queue.
+          if (isMergedBlockAddressRemote(address)) {
+            // Fallback for all the pending fetch requests
+            val pendingShuffleChunks = iterator.removePendingChunks(shuffleChunkId, address)
+            if (pendingShuffleChunks.nonEmpty) {
+              pendingShuffleChunks.foreach { pendingBlockId =>
+                logInfo(s"Falling back immediately for merged block $pendingBlockId")
+                val bitmapOfPendingChunk: RoaringBitmap = chunksMetaMap.remove(pendingBlockId)
+                assert(bitmapOfPendingChunk != null)
+                chunkBitmap.or(bitmapOfPendingChunk)
+              }
+              // These blocks were added to numBlocksToFetch so we increment numBlocksProcessed
+              blocksProcessed += pendingShuffleChunks.size
+            }
+          }
+          mapOutputTracker.getMapSizesForMergeResult(
+            shuffleChunkId.shuffleId, shuffleChunkId.reduceId, chunkBitmap)
+      }
+    iterator.fetchFallbackBlocks(fallbackBlocksByAddr)
```

Comment:
`fallback` makes it clear here that it is going fetch all the original blocks which were part of the merged block but failed. Just hesitant to use the terminology `fetchUnMergedBlocks` because these were merged. 
However, don't have a strong opinion about this.

Reply:
My thought is:

`fallback` is a strategy here and we don't have a type of block named `fallback block`...

Maybe, rename `fetchFallbackBlocks` to `fallbackFetch`?

And rename `fallbackBlocksByAddr` to `originalBlocksByAddr` or `individulBlocksByAddr` or something else better if you don't like `unMergedBlocksByAddr`.

Reply:
done

Reply:
done